### PR TITLE
Product SKU: Adds support for color, typography, and spacing

### PR DESCRIPTION
**Pull request by imanish003**
This Pull Request enhances the styling capabilities of the Product SKU block by introducing additional support for color, typography & spacing. This improvement provides users with more flexibility when customizing the appearance of the Product SKU block.

Here is what Product SKU block supports now:

**Color:**
- Text color
- Background color

**Typography:**
- Font family 
- Font size
- Font Appearance
- Line height
- Letter spacing 
- Text decoration 
- Letter Case

**Spacing:**
- Margin
- Padding

Part of #8691 & #7954

#### Test using blocks sidebar
- Create a new page or post
- Add the "Products" block to your post.
- Add Product SKU block as an inner block if not included already
- Select the Product SKU block and open the block settings in the right sidebar.
- Customize the block using the new styling options available:
   - Change the text color, background color, font family, font size, font appearance, line height, letter spacing, text decoration, letter case, margin, and padding under the appropriate settings sections.
- Verify that the styling changes are immediately visible and applied correctly within the editor itself.
- Save the post, then view it on the front-end to ensure the styling changes are visible and applied correctly.

#### Test using Global Styles
- Open the Global Styles sidebar by clicking the "Styles" button at the top right corner of the editor.

https://user-images.githubusercontent.com/16707866/228781997-24a9a29f-4be7-46d4-b1b3-41520692d86d.mov

- In the Global Styles sidebar, locate the "Product SKU" block.
- Customize the block using the new styling options available under the "Product SKU" block in the Global Styles sidebar:
   - Change the text color, background color, font family, font size, font appearance, line height, letter spacing, text decoration, letter case, margin, and padding under the appropriate settings sections.
- Verify that these applied correctly within editor & on frontend:
   - Create a new post
   - Add Products block
   - Add Product SKU block as an inner block if not included already
   - Check the Product SKU block in both the editor and on the front-end to ensure that the changes made through the Global styles UI are accurately applied:
       - In the editor, inspect Product SKU block & ensure it matches the values set in the Global styles UI.
       - Save the post & publish it. On frontend, inspect Product SKU block & ensure it matches the values set in the Global styles UI.

> Also, verify the changes in All Products block, which also uses Product SKU block.

* [ ] Do not include in the Testing Notes 

### WooCommerce Visibility

* [ ] WooCommerce Core
* [X] Feature plugin
* [ ] Experimental

### Changelog

> Enhance Product SKU block styling capabilities with additional support for color, typography & spacing

### CODE REVIEW

**Comment by imanish003** [comment]
@Aljullu, I am also adding you as reviewer because this PR adds `isDescendantOfAllProducts` flag for all products elements.
I noticed that product elements require special handling when used inside All products block. IMO this flag will be very useful in adding any new functionalities to product elements while still supporting All Products block.

I also considered creating a recursive function that checks block parents and their ancestors. Something like this:
```
function isDescendantOfAllProductsBlock(blockClientId, getBlockParents) {
  const parentClientIds = getBlockParents(blockClientId);

  if (parentClientIds.length === 0) {
    return false;
  }

  for (const parentId of parentClientIds) {
    const parentBlock = select('core/block-editor').getBlock(parentId);
    if (parentBlock.name === 'woocommerce/all-products') {
      return true;
    }

    if (isDescendantOfAllProductsBlock(parentId, getBlockParents)) {
      return true;
    }
  }

  return false;
}
```

But I dropped this idea because I thought this would have performance impact. Please let me know your thoughts about it 🙂

File: assets/js/atomic/blocks/product-elements/sku/supports.ts
Comment (Aljullu) [inline assets/js/atomic/blocks/product-elements/sku/supports.ts]: Heads-up that this line introduced an ESLint warning.

File: assets/js/atomic/blocks/product-elements/sku/block.tsx
Comment (kmanijak) [inline assets/js/atomic/blocks/product-elements/sku/block.tsx:16]: I'm wondering if there's any other way of achieving that or can the hook be used on the Editor side?

The reason I'm asking is these hooks have a dependency on `@wordpress/block-editor` which has a dependency on `@wordpress/components` ([issue](https://github.com/woocommerce/woocommerce-blocks/issues/8071)) and neither of them should be used on the frontend because they're not designed for this and have a performance impact.

To be fair - it's not a blocker, because those hooks are already widely used on a frontend and given the `@wordpress` dependencies are externalized, there's no additional impact by yet another use of hooks (no matter the number of usages, the dependency is loaded just once). Having said that, I'm just raising this for awareness and maybe there's an easy way of avoiding adding another use of that (don't know that 😅 ).

File: assets/js/atomic/blocks/product-elements/sku/edit.tsx
Comment (kmanijak) [inline assets/js/atomic/blocks/product-elements/sku/edit.tsx:23]: Seems `wp-block-woocommerce-product-sku` is duplicated in a [block here](https://github.com/woocommerce/woocommerce-blocks/pull/8913/files#diff-60f7e4bda641c87212e5a65b413e1c794870354e6431235268e5794f784c893fR80). Can one of occurrances be removed?

File: assets/js/atomic/blocks/product-elements/sku/supports.ts
Comment (Aljullu) [inline assets/js/atomic/blocks/product-elements/sku/supports.ts]: Is `__experimentalSelector` required? I tried removing it and everything seemed to keep working as expected. :thinking:

File: assets/js/atomic/blocks/product-elements/sku/supports.ts
Comment (Aljullu) [inline assets/js/atomic/blocks/product-elements/sku/supports.ts]: If we can remove `__experimentalSelector`, then we could make these styles available in WC core as well, so only the experimental ones would be gated to the feature plugin. :slightly_smiling_face:

**Review by Aljullu — COMMENTED** [review]
It's testing perfectly and code looks good to me. Nice improvement to the Product SKU block. :clap: 

Not sure if it's the expected behavior, but font-weight is only applied to the `SKU:` prefix, but not to the value:

imatge

> @Aljullu, I am also adding you as reviewer because this PR adds `isDescendantOfAllProducts` flag for all products elements.

I see we have similar flag for other parent blocks, so it makes sense to me. :slightly_smiling_face:

**Review by kmanijak — COMMENTED** [review]
Thanks for taking care of this! Things I spotted while testing:


1. Little caveat (edge case I think) is the text is wrapped in the Editor while it may escape the container in the frontend. Below is the example in All Products, but it behaves the same in Products block.

| Editor | Frontend |
|---|---|
| image| image |

2. Letter case is not reflected in the Editor but it works ok in Frontend while used in GLobal Styles. Below you can see it's set to lowercase in Editor, but it's kept uppercase in there.

| Editor | Frontend |
|---|---|
| image| image|

Everything else looks good! I left some minor comments otherwise.

Also, the above may require the changes in Gutenberg, and in such case, I don't think they should block this PR from being merged 🙌

File: assets/js/atomic/blocks/product-elements/sku/supports.ts
Comment (imanish003) [reply assets/js/atomic/blocks/product-elements/sku/supports.ts]: Good catch! I think we can remove `__experimentalSelector` flag.

File: assets/js/atomic/blocks/product-elements/sku/supports.ts
Comment (imanish003) [reply assets/js/atomic/blocks/product-elements/sku/supports.ts]: That makes sense. I will make the necessary changes 🤝

File: assets/js/atomic/blocks/product-elements/sku/supports.ts
Comment (imanish003) [reply assets/js/atomic/blocks/product-elements/sku/supports.ts]: ESLint warning is present wherever we use __experimentalGetSpacingClassesAndStyles, therefore I thought it was acceptable. However, I will add the following lines to ignore the warning going forward 🙂
```ts
// @ts-expect-error We check if this exists before using it.
// eslint-disable-next-line @wordpress/no-unsafe-wp-apis
```

**Comment by imanish003** [comment]
@Aljullu 
> Not sure if it's the expected behavior, but font-weight is only applied to the SKU: prefix, but not to the value:

Actually, we have value text in a strong tag:
```
<strong>{ value }</strong>
```

I thought we wanna keep `strong` tag. I don't have a strong opinion about it, if you want, I can remove the strong tag. 🙂

> I see we have similar flag for other parent blocks, so it makes sense to me. 

Excellent, this will simplify the process of keeping product elements compatible with the All products block! 🙌🏻

File: assets/js/atomic/blocks/product-elements/sku/edit.tsx
Comment (imanish003) [reply assets/js/atomic/blocks/product-elements/sku/edit.tsx:23]: I apologize, but I am having trouble understanding your comment. Are you suggesting that `wc-block-components-product-sku` is equivalent to `wp-block-woocommerce-product-sku`? 🤔

Comment (kmanijak) [reply assets/js/atomic/blocks/product-elements/sku/edit.tsx:23]: > Are you suggesting that wc-block-components-product-sku is equivalent to wp-block-woocommerce-product-sku? 🤔

Sorry I wasn't clear enough. What I mean is:

1. Here you provide a class name `wp-block-woocommerce-product-sku`, which is assigned to the `div` block container: [here](https://github.com/woocommerce/woocommerce-blocks/pull/8913/files#diff-273af84aa585a8e5a5a1446b8e97d85242eca59f5bf0696f3111b6627794ebaeR40)
2. The same class `wp-block-woocommerce-product-sku` is assigned to the Preview component [in here](https://github.com/woocommerce/woocommerce-blocks/pull/8913/files#diff-60f7e4bda641c87212e5a65b413e1c794870354e6431235268e5794f784c893fR80).

But I think I see what's happening, these are mutually exclusive, right? 🤔

Comment (imanish003) [reply assets/js/atomic/blocks/product-elements/sku/edit.tsx:23]: Thanks for the explanation. Now I can understand what you mean.

Actually, [here in block.tsx](https://github.com/woocommerce/woocommerce-blocks/blob/9e10039b4471566084ed3b99b808db92e3b713eb/assets/js/atomic/blocks/product-elements/sku/block.tsx#L77-L86), we are adding a class name for `All products`. Because classname added in [edit.tsx](https://github.com/woocommerce/woocommerce-blocks/blob/9e10039b4471566084ed3b99b808db92e3b713eb/assets/js/atomic/blocks/product-elements/sku/edit.tsx#L23) won't be added to `All products` blocks.

In summary, the class name in block.tsx will apply to the All products block, while the class name in edit.tsx will apply everywhere else. I hope that makes sense. 

Although I think this kind of handling shouldn't be required, but to support product elements in `All Products`, I added this 🙂

File: assets/js/atomic/blocks/product-elements/sku/block.tsx
Comment (imanish003) [reply assets/js/atomic/blocks/product-elements/sku/block.tsx:16]: TBH, I was thinking about writing a P2 post about it. Ideally, we shouldn't need useColorProps, useSpacingProps, useTypographyProps hooks. 

Instead [here](https://github.com/woocommerce/woocommerce-blocks/blob/9e10039b4471566084ed3b99b808db92e3b713eb/assets/js/atomic/blocks/product-elements/save.js#L15) in save.js, we should return following instead:
```
export const Save = ( { attributes }: Props ): JSX.Element => {
	return (
		<div
			{ ...useBlockProps.save( {
				className: classnames( 'is-loading', attributes.className ),
			} ) }
		/>
	);
};
```

AFAIK `useBlockProps` hooks will generate `className` & `styles`, which we generate using these hooks(useColorProps, useSpacingProps, useTypographyProps). 
But I believe as we need to support product elements in `All products` block, we are using these hooks. 

@Aljullu Is my understanding correct? 🤔

Comment (Aljullu) [reply assets/js/atomic/blocks/product-elements/sku/block.tsx:16]: Just to clarify, these hooks are only loaded in the frontend in the _All Products_ block, correct? In that case, I would go ahead with this. If I'm not wrong, we have some other product element blocks that depend on those hooks as well. And sooner or later we will replace the _All Products_ block with the _Product_ block.

> I was thinking about writing a P2 post about it.

That would be super helpful. :bow:

**Review by Aljullu — APPROVED** [review]
Looks good on my end, so approving. Leaving @kmanijak to do the final review on the items he brought up to the discussion.

> I thought we wanna keep `strong` tag. I don't have a strong opinion about it, if you want, I can remove the strong tag. 🙂

I don't have a specific preference either. I would say let's keep it as-is now. At some point it might be good to get design feedback on it, though.

**Comment by imanish003** [comment]
@Aljullu I appreciate your valuable input 🙏  I agree with you; we can keep the strong tag for now, and as you suggested, we can seek design feedback at a later stage. I'll wait for @kmanijak final review and address any concerns he may have before merging this PR 🙌🏻

File: assets/js/atomic/blocks/product-elements/sku/block.tsx
Comment (kmanijak) [reply assets/js/atomic/blocks/product-elements/sku/block.tsx:16]: > If I'm not wrong, we have some other product element blocks that depend on those hooks as well.

That's true and as I wrote earlier "...I'm just raising this for awareness...". So I'm resolving the thread 👍

**Review by kmanijak — APPROVED** [review]
> I'll wait for @kmanijak final review and address any concerns he may have before merging this PR 🙌🏻

Excellent work! I double checked the previously raised glitches and smoke tested - all works as expected. Great work @imanish003! 💪

**Comment by opr** [comment]
Hey @imanish003 when editing the case of the SKU block, it also affects the SKU itself. Is this intended? Would we prefer to display the SKU exactly as the merchant entered in the product page?

**Comment by imanish003** [comment]
> Is this intended? Would we prefer to display the SKU exactly as the merchant entered in the product page?

This is intended as per the current implementation. I understand your concern about displaying the SKU as the merchant entered it. 
However, I'm not entirely sure whether we should display the SKU exactly as the merchant entered it because it is still possible to customize it the way merchant wants it using additional supports which were added in this PR. 

Perhaps we could get input from @shaunandrews on this matter? 🙂

**Pull request merged**
